odl/space/weighting.py

```diff
@@ -0,0 +1,1047 @@
+            else:
+                inner_fstr += '\n'
+
+        inner_str = inner_fstr.format(shape=self.matrix.shape, fmt=fmt,
+                                      nnz=nnz, ex=self.exponent,
+                                      matrix=self.matrix)
+        return '{}({})'.format(self.__class__.__name__, inner_str)
+
+    def __str__(self):
+        """Return ``str(self)``."""
+        if self.exponent == 2.0:
+            return 'Weighting: matrix =\n{}'.format(self.matrix)
+        else:
+            return 'Weighting: p = {}, matrix =\n{}'.format(self.exponent,
+                                                            self.matrix)
+
+
+class VectorWeighting(WeightingBase):
+
+    """Weighting of a space by a vector.
+
+    The exact definition of the weighted inner product, norm and
+    distance functions depend on the concrete space.
+
+    The vector may only have positive entries, otherwise it does not
+    define an inner product or norm, respectively. This is not checked
+    during initialization.
+    """
+
+    def __init__(self, vector, impl, exponent=2.0, dist_using_inner=False):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        vector : `array-like`, one-dim.
+            Weighting vector of the inner product
+        impl : `str`
+            Specifier for the implementation backend
+        exponent : positive `float`
+            Exponent of the norm. For values other than 2.0, the inner
+            product is not defined.
+            If ``matrix`` is a sparse matrix, only 1.0, 2.0 and ``inf``
+            are allowed.
+        dist_using_inner : `bool`, optional
+            Calculate `dist` using the following formula::
+
+                ||x - y||^2 = ||x||^2 + ||y||^2 - 2 * Re <x, y>
+
+            This avoids the creation of new arrays and is thus faster
+            for large arrays. On the downside, it will not evaluate to
+            exactly zero for equal (but not identical) ``x`` and ``y``.
+
+            This option can only be used if ``exponent`` is 2.0.
+        """
+        super().__init__(impl=impl, exponent=exponent,
+                         dist_using_inner=dist_using_inner)
+
+        if isinstance(vector, FnBaseVector):
+            self._vector = vector
+        else:
+            self._vector = np.asarray(vector)
+
+        if self.vector.dtype == object:
+            raise ValueError('invalid vector {}.'.format(vector))
+        elif self.vector.ndim != 1:
+            raise ValueError('vector {} is {}-dimensional instead of '
+                             '1-dimensional.'
+                             ''.format(vector, self._vector.ndim))
+
+    @property
+    def vector(self):
+        """Weighting vector of this inner product."""
+        return self._vector
+
+    def vector_is_valid(self):
+        """Test if the vector is a valid weight, i.e. positive."""
+        return np.all(np.greater(self.vector, 0))
```

Comment:
Not remembering your own suggestions? :-)

This version works with `CudaFnVector`.

The method is not used externally at the moment, but I see these validation functions as a debugging interface. Maybe for vectors, it's simple, but for matrices, it gets more involved.